drud/ddev
# Rethink hooks and tasks, for #1372, fixes #1321, fixes #1071, fixes #1038

## The Problem/Issue/Bug:

Big picture is #1372: Hook System Overhaul
Fixes #1321 "Add composer to hook commands"
Fixes #1071 "Provide hooks for other commands"
Fixes #1038 - Specify target container for exec command

## How this PR Solves The Problem:

## TODO:
- [x] Allow configuration of the service=container that a task is to run on
- [x] Add composer exec task
- [x] Add hooks for pre/post-composer, pre/post-stop, config, describe, exec, list, pause, pull, restart?, snapshot, restore-snapshot
- [x] Update docs to show new hook configuration options
- [x] Add composer task to docs
- [x] Change verbiage to use "task"
- [x] Remove the poor validation of hooks in favor of actual creation and use of them?

## Manual Testing Instructions:

* Add pre- and post- hooks for some of the hooks. post-composer and post-stop exec hooks maybe? Execute the command and verify the behavior. 
* Use the composer task in a hook, perhaps a post-composer hook that does a composer task
* Create an exec task that executes on the db or dba containers

## Automated Testing Overview:

- [x] Test composer task
- [x] post-composer hook
- [x] Other pre- and post- hooks

## Related Issue Link(s):

## Release/Deployment notes:

* The request for commands here needs to be spun off into a different issue, with the comments required... Or else leave this here and change the title.
